### PR TITLE
Add a fuzz! version that doesn't hook panics

### LINKED ISSUE #150: Can't fuzz a crate that catches panic
My rawloader crate catches panic on file decode to just ignore broken files without having to rewrite all the bounds checks that rust already provides. I'd still like to fuzz it to catch hangs and other causes of crashes. It doesn't seem to be possible right now though because all panics are turned into aborts. Is there a way to avoid that?

### PR DESCRIPTION
**Pull request by pedrocr**
If your fuzzing code catches panics somewhere inside its code the hook would turn those into crashes. Allow disabling the hook by adding a fuzz_nohook! macro.

Even in empty targets the hook causes variability in the execution of the code somehow. That reduces the stability of the fuzzing making it less effective.

Fixes #150

### CODE REVIEW

**Comment by Shnatsel** [comment]
I'm interested in this feature as well - I'm planning a large-scale fuzzing run where we have little info about fuzz targets, so panics (i.e. indicating misuse of the API) is acceptable and expected, but segfaults are not.

File: src/lib.rs [hunk 1]
```diff
@@ -135,17 +135,19 @@ where
...
+      }));
+    }
 
     let mut input = vec![];
 
     // initialize forkserver there
     unsafe { __afl_manual_init() };
 
-    while unsafe { __afl_persistent_loop(1000) } != 0 {
+    while unsafe { __afl_persistent_loop(1) } != 0 {
```
Comment (frewsxcv) [inline src/lib.rs]: @pedrocr Do you remember why this change was made?

Comment (pedrocr) [reply src/lib.rs]: Unfortunately I forget why that was needed. Not sure if it just didn't work without that or if it was just because of performance. I can't find any docs for that function either to even know what the parameter does.

Comment (pedrocr) [reply src/lib.rs]: I've just tested it and it works without this. I've pushed a version that rebased to master and doesn't have this change. The diff looks pretty clean now.

Comment (Shnatsel) [reply src/lib.rs]: I don't know for sure, but I suspect that the parameter is how frequently the process is restarted during in-process fuzzing, in number of executions before restart. So if you set the parameter to 1 it should be noticeably slower, almost equivalent to the performance of stdin mode.

File: src/lib.rs [hunk 2]
```diff
@@ -195,13 +197,62 @@ where
...
+///     if data[2] != b'e' {return}
+///     if data[3] != b'r' {return}
+///     if data[4] != b't' {return}
+///     if data[5] != b'y' {return}
+///     panic!("BOOM")
+/// });
+/// # }
+/// ```
+#[macro_export]
+macro_rules! fuzz_nohook {
```
Comment (Shnatsel) [inline src/lib.rs]: `fuzz_nohook!` is not a particularly readable name. How about `fuzz_no_catch_panic!` or something like that?

Comment (pedrocr) [reply src/lib.rs]: Fine by me. Let me know what name you prefer.

**Comment by Shnatsel** [comment]
Correction: this PR should actually be unrelated to what I'm trying to accomplish. The macro from this PR would not make the panic register as a crash, but still let panics terminate the process. My use case calls for catching panics and ignoring them without letting them terminate the process, since terminating the process would slow down fuzzing by a factor of up to 10x.

I have (belatedly) realized that ignoring panics can actually be achieved by catching the panic inside the closure passed to `fuzz!` - unless there is some mechanism in Rust that prevents this that I'm not aware of. I'll create a prototype with a custom panic catcher and report back.

**Comment by pedrocr** [comment]
@Shnatsel you need both this PR and to catch panic in your code. `rawloader` catches panic but without this code the panics get turned into crashes even if they are caught in your code.

**Comment by frewsxcv** [comment]
Thanks!
